guides/source/threading_and_code_execution.md

```diff
@@ -0,0 +1,262 @@
+**DO NOT READ THIS FILE ON GITHUB, GUIDES ARE PUBLISHED ON http://guides.rubyonrails.org.**
+
+Threading and Code Execution in Rails
+=====================================
+
+After reading this guide, you will know:
+
+* What code Rails will automatically execute concurrently
+* How to integrate manual concurrency with Rails internals
+* How to wrap all application code
+* How to affect application reloading
+
+--------------------------------------------------------------------------------
+
+Automatic Concurrency
+---------------------
+
+Rails automatically allows various operations to be performed at the same time.
+
+When using a threaded web server, such as the default Puma, multiple HTTP
+requests will be served simultaneously, with each request provided its own
+controller instance.
+
+Threaded Active Job adapters, including the built-in Async, will likewise
+execute several jobs at the same time. Action Cable channels are managed this
+way too.
+
+These mechanisms all involve multiple threads, each managing work for a unique
+instance of some object (controller, job, channel), while sharing the global
+process space (such as classes and their configurations, and global variables).
+As long as your code doesn't modify any of those shared things, it can mostly
+ignore that other threads exist.
+
+The rest of this guide describes the mechanisms Rails uses to make it "mostly
+ignorable", and how extensions and applications with special needs can use them.
+
+Executor
+--------
+
+The Rails Executor separates application code from framework code: any time the
+framework invokes code you've written in your application, it will be wrapped by
+the Executor.
+
+The Executor consists of two callbacks: `to_run` and `to_complete`. The Run
+callback is called before the application code, and the Complete callback is
+called after.
+
+### Default callbacks
+
+In a default Rails application, the Executor callbacks are used to:
+
+* track which threads are in safe positions for autoloading and reloading
+* enable and disable the Active Record query cache
+* return acquired Active Record connections to the pool
+* constrain internal cache lifetimes
+
```

Comment:
This Guide is great, thanks. Can you reference or link to where these callbacks are set up for default Rails app?

Reply:
I don't think there's any precedent for that degree of source-code tourism in guides; the Autoloading and Reloading Constants guide is a very deep dive, but it still doesn't get that close to the implementation.

Apart from the risk of getting a reader bogged down in implementation detail (e.g. how our callbacks work, as opposed to how they're registered), I think the biggest worry with such links would be that they'll rot fairly quickly: these are couple-of-line instances in different places throughout the framework.

[Like](https://github.com/matthewd/rails/blob/b5cd7d5b511361062f67ebad45ab38393970fc47/railties/lib/rails/application/finisher.rb#L120) [seriously](https://github.com/matthewd/rails/blob/b5cd7d5b511361062f67ebad45ab38393970fc47/railties/lib/rails/application/finisher.rb#L155-L185) [there](https://github.com/matthewd/rails/blob/b5cd7d5b511361062f67ebad45ab38393970fc47/railties/lib/rails/application/default_middleware_stack.rb#L40-L53) [are](https://github.com/matthewd/rails/blob/b5cd7d5b511361062f67ebad45ab38393970fc47/actioncable/lib/action_cable/engine.rb#L56-L75) [just](https://github.com/matthewd/rails/blob/b5cd7d5b511361062f67ebad45ab38393970fc47/activejob/lib/active_job/railtie.rb#L26-L30) [so](https://github.com/matthewd/rails/blob/b5cd7d5b511361062f67ebad45ab38393970fc47/activerecord/lib/active_record/railtie.rb#L157-L162) [many](https://github.com/matthewd/rails/blob/b5cd7d5b511361062f67ebad45ab38393970fc47/activerecord/lib/active_record/query_cache.rb#L28-L47) [of](https://github.com/matthewd/rails/blob/b5cd7d5b511361062f67ebad45ab38393970fc47/actionview/lib/action_view/railtie.rb#L54) [them](https://github.com/matthewd/rails/blob/b5cd7d5b511361062f67ebad45ab38393970fc47/activesupport/lib/active_support/railtie.rb#L20-L22)[.](https://github.com/matthewd/rails/blob/b5cd7d5b511361062f67ebad45ab38393970fc47/railties/lib/rails/console/app.rb#L34)